File: library/src/androidTest/java/com/raizlabs/android/dbflow/test/sql/TriggerTest.java
package com.raizlabs.android.dbflow.test.sql;

import com.raizlabs.android.dbflow.sql.builder.Condition;
import com.raizlabs.android.dbflow.sql.language.Delete;
import com.raizlabs.android.dbflow.sql.language.Select;
import com.raizlabs.android.dbflow.sql.language.Update;
import com.raizlabs.android.dbflow.sql.language.Where;
import com.raizlabs.android.dbflow.sql.trigger.CompletedTrigger;
import com.raizlabs.android.dbflow.sql.trigger.Trigger;
import com.raizlabs.android.dbflow.test.FlowTestCase;
import com.raizlabs.android.dbflow.test.structure.TestModel1;
import com.raizlabs.android.dbflow.test.structure.TestModel1$Table;

/**
 * Description:
 */
public class TriggerTest extends FlowTestCase {


    public void testTriggerLanguage() {
        Where<TestModel1> logic = new Update().table(TestModel1.class)
                .set(Condition.column(TestModel1$Table.NAME).is("Jason"))
                .where(Condition.column(TestModel1$Table.NAME).is("Jason2"));
        String trigger = new Trigger<ConditionModel>("MyTrigger")
                .after().insert(ConditionModel.class).begin(logic).getQuery();
        assertEquals("CREATE TRIGGER IF NOT EXISTS MyTrigger  AFTER INSERT ON ConditionModel " +
                "\nBEGIN" +
                    "\n" + logic.getQuery() +";" +
                "\nEND", trigger.trim());

        trigger = new Trigger<ConditionModel>("MyTrigger2")
                .before().update(ConditionModel.class, ConditionModel$Table.NAME).begin(logic).getQuery();
        assertEquals("CREATE TRIGGER IF NOT EXISTS MyTrigger2  BEFORE UPDATE OF name ON ConditionModel " +
                "\nBEGIN" +
                "\n" + logic.getQuery() + ";" +
                "\nEND", trigger.trim());
    }

    public void testTriggerFunctions() {
        Delete.tables(TestUpdateModel.class, ConditionModel.class);

        CompletedTrigger<ConditionModel> trigger = new Trigger<ConditionModel>("TestTrigger")
                    .after().insert(ConditionModel.class).begin(new Update().table(TestUpdateModel.class)
                        .set(Condition.column(TestUpdateModel$Table.VALUE).is("Fired")));

        TestUpdateModel model = new TestUpdateModel();
        model.name = "Test";
        model.value = "NotFired";
        model.save();

        trigger.enable();

        ConditionModel conditionModel = new ConditionModel();
        conditionModel.name = "Test";
        conditionModel.fraction = 0.6d;
        conditionModel.insert();

        model = new Select().from(TestUpdateModel.class)
                .where(Condition.column(TestUpdateModel$Table.NAME).is("Test")).querySingle();
        assertEquals(model.value, "Fired");

        trigger.disable();
    }

}

File: dbflow-tests/src/test/java/com/raizlabs/dbflow5/dbflow/sql/language/TriggerTest.kt
package com.raizlabs.dbflow5.dbflow.sql.language

import com.raizlabs.dbflow5.dbflow.BaseUnitTest
import com.raizlabs.dbflow5.dbflow.assertEquals
import com.raizlabs.dbflow5.config.databaseForTable
import com.raizlabs.dbflow5.dbflow.models.SimpleModel
import com.raizlabs.dbflow5.dbflow.models.SimpleModel_Table.name
import com.raizlabs.dbflow5.dbflow.models.TwoColumnModel
import com.raizlabs.dbflow5.dbflow.models.TwoColumnModel_Table.id
import com.raizlabs.dbflow5.query.NameAlias
import com.raizlabs.dbflow5.query.begin
import com.raizlabs.dbflow5.query.cast
import com.raizlabs.dbflow5.query.columnValues
import com.raizlabs.dbflow5.query.createTrigger
import com.raizlabs.dbflow5.query.insert
import com.raizlabs.dbflow5.query.insertOn
import com.raizlabs.dbflow5.query.property.property
import com.raizlabs.dbflow5.query.updateOn
import com.raizlabs.dbflow5.sql.SQLiteType
import com.raizlabs.dbflow5.structure.insert
import org.junit.Assert.assertNotNull
import org.junit.Test

class TriggerTest : BaseUnitTest() {

    @Test
    fun validateBasicTrigger() {
        databaseForTable<SimpleModel> {
            assertEquals("CREATE TRIGGER IF NOT EXISTS `MyTrigger` AFTER INSERT ON `SimpleModel` " +
                    "\nBEGIN" +
                    "\nINSERT INTO `TwoColumnModel`(`name`) VALUES(`new`.`name`);" +
                    "\nEND",
                    createTrigger("MyTrigger").after() insertOn SimpleModel::class begin
                            insert(TwoColumnModel::class).columnValues(name to NameAlias.ofTable("new", "name")))
        }
    }

    @Test
    fun validateUpdateTriggerMultiline() {
        databaseForTable<SimpleModel> {
            assertEquals("CREATE TEMP TRIGGER IF NOT EXISTS `MyTrigger` BEFORE UPDATE ON `SimpleModel` " +
                    "\nBEGIN" +
                    "\nINSERT INTO `TwoColumnModel`(`name`) VALUES(`new`.`name`);" +
                    "\nINSERT INTO `TwoColumnModel`(`id`) VALUES(CAST(`new`.`name` AS INTEGER));" +
                    "\nEND",
                    createTrigger("MyTrigger").temporary().before() updateOn SimpleModel::class begin
                            insert(TwoColumnModel::class).columnValues(name to NameAlias.ofTable("new", "name")) and
                            insert(TwoColumnModel::class)
                                    .columnValues(id to cast(NameAlias.ofTable("new", "name").property)
                                            .`as`(SQLiteType.INTEGER)))
        }
    }

    @Test
    fun validateTriggerWorks() {
        databaseForTable<SimpleModel> {
            val trigger = createTrigger("MyTrigger").after() insertOn SimpleModel::class begin
                    insert(TwoColumnModel::class).columnValues(name to NameAlias.ofTable("new", "name"))
            trigger.enable()
            SimpleModel("Test").insert()

            val result = select from TwoColumnModel::class where (name eq "Test")
            assertNotNull(result)
        }
    }
}